### PR TITLE
Implement PermissionService API changes

### PR DESCRIPTION
**Pull request by lucko**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1524) | **Common**

Implements SpongePowered/SpongeAPI#1524

### CODE REVIEW

File: src/main/java/org/spongepowered/common/service/permission/OpLevelCollection.java [hunk 1]
```diff
@@ -122,6 +129,16 @@ public SubjectCollection getContainingCollection() {
         }
 
         @Override
+        public PermissionService getService() {
+            return service;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/OpLevelCollection.java]: `this.`

File: src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    public SpongeSubjectReference(SpongePermissionService service, String collectionId, String subjectId) {
+        this.service = service;
+        this.collectionId = collectionId;
+        this.subjectId = subjectId;
+    }
+
+    @Override
+    public String getCollectionIdentifier() {
+        return collectionId;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java]: `this.`

File: src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public String getSubjectIdentifier() {
+        return subjectId;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java]: `this.`

File: src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public synchronized CompletableFuture<Subject> resolve() {
+        // lazily load
+        if (cache == null) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java]: `this.`

File: src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            cache = service.get(collectionId).get(subjectId);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java]: `this.`

File: src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        return CompletableFuture.completedFuture(cache);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/permission/SpongeSubjectReference.java]: `this.`

**Review by kashike — CHANGES_REQUESTED** [review]
Please go through all your changes and ensure you have `this.`.

**Comment by kashike** [comment]
:eyes:

**Comment by lucko** [comment]
Yeah, lost all my changes trying to rebase. Didn't like that I was targeting a different submodule, and I lost everything trying to merge with upstream

Really don't have it in me to re-do it. I spent hours and hours trying to work out how to even target a different repository in the initial PR.

**Comment by gabizou** [comment]
You can recover previous commits (in this case references) with git ref-log. You should be able to revert the branch to the previous reference before you rebased.

**Comment by lucko** [comment]
I had it cloned inside of SpongeForge, so at least got the commit.

https://gist.github.com/lucko/e23e18301021eb849816ffbd08e2210f

Not sure if that's useful if anyone else wants to try merging it. It was tested and working, aside from some codestyle issues.

**Comment by kashike** [comment]
We've adopted your changed into a branch: https://github.com/SpongePowered/SpongeCommon/compare/lucko/permission

Thanks for your work on this, it is very much appreciated.

**Pull request closed without merging**